File: frontend/src/lib/api/authConfig.ts
/**
 * API client for authentication configuration management.
 */
import { axiosInstance } from '../axios';

export interface AuthConfigResponse {
  id: number;
  uuid: string;
  config_key: string;
  config_value: string | null;
  is_sensitive: boolean;
  category: string;
  data_type: string;
  description: string | null;
  requires_restart: boolean;
  created_at: string;
  updated_at: string;
}

export interface AuthConfigAuditResponse {
  id: number;
  uuid: string;
  config_key: string;
  old_value: string | null;
  new_value: string | null;
  change_type: string;
  ip_address: string | null;
  created_at: string;
}

export interface AuthMethodTestResponse {
  success: boolean;
  message: string;
  details?: Record<string, any>;
}

export interface LDAPConfig {
  ldap_enabled: boolean;
  ldap_server: string;
  ldap_port: number;
  ldap_use_ssl: boolean;
  ldap_use_tls: boolean;
  ldap_bind_dn: string;
  ldap_bind_password?: string;
  ldap_search_base: string;
  ldap_username_attr: string;
  ldap_email_attr: string;
  ldap_name_attr: string;
  ldap_user_search_filter: string;
  ldap_timeout: number;
  ldap_admin_users: string;
  ldap_admin_groups: string;
  ldap_user_groups: string;
  ldap_recursive_groups: boolean;
  ldap_group_attr: string;
}

export interface KeycloakConfig {
  keycloak_enabled: boolean;
  keycloak_server_url: string;
  keycloak_internal_url: string;
  keycloak_realm: string;
  keycloak_client_id: string;
  keycloak_client_secret?: string;
  keycloak_callback_url: string;
  keycloak_admin_role: string;
  keycloak_timeout: number;
  keycloak_verify_audience: boolean;
  keycloak_audience: string;
  keycloak_use_pkce: boolean;
  keycloak_verify_issuer: boolean;
}

export interface PKIConfig {
  pki_enabled: boolean;
  pki_ca_cert_path: string;
  pki_verify_revocation: boolean;
  pki_cert_header: string;
  pki_cert_dn_header: string;
  pki_admin_dns: string;
  pki_ocsp_timeout_seconds: number;
  pki_crl_cache_seconds: number;
  pki_revocation_soft_fail: boolean;
  pki_trusted_proxies: string;
  pki_mode: string;
  pki_allow_password_fallback: boolean;
}

export interface SessionConfig {
  jwt_access_token_expire_minutes: number;
  jwt_refresh_token_expire_days: number;
  session_idle_timeout_minutes: number;
  session_absolute_timeout_minutes: number;
  max_concurrent_sessions: number;
  concurrent_session_policy: string;
}

export class AuthConfigApi {
  static async getAllConfigs(): Promise<Record<string, AuthConfigResponse[]>> {
    const response = await axiosInstance.get('/admin/auth-config');
    return response.data;
  }

  static async getConfigByCategory(category: string): Promise<Record<string, any>> {
    const response = await axiosInstance.get(`/admin/auth-config/${category}`);
    return response.data;
  }

  static async updateCategory(category: string, config: Record<string, any>): Promise<void> {
    await axiosInstance.put(`/admin/auth-config/${category}`, config);
  }

  static async testConnection(
    category: string,
    config: Record<string, any>
  ): Promise<AuthMethodTestResponse> {
    const response = await axiosInstance.post(`/admin/auth-config/${category}/test`, config);
    return response.data;
  }

  static async getAuditLog(
    category: string,
    limit: number = 100
  ): Promise<AuthConfigAuditResponse[]> {
    const response = await axiosInstance.get(`/admin/auth-config/audit/${category}`, {
      params: { limit },
    });
    return response.data;
  }

  static async migrateFromEnv(): Promise<{ migrated_count: number }> {
    const response = await axiosInstance.post('/admin/auth-config/migrate');
    return response.data;
  }
}

File: backend/app/schemas/auth_config.py
"""Pydantic schemas for authentication configuration."""

from datetime import datetime
from typing import Any
from typing import Optional

from pydantic import BaseModel
from pydantic import Field


class AuthConfigBase(BaseModel):
    """Base schema for authentication configuration."""

    config_key: str = Field(..., max_length=100)
    config_value: Optional[str] = None
    is_sensitive: bool = False
    category: str = Field(..., max_length=50)
    data_type: str = Field(default="string", max_length=20)
    description: Optional[str] = None
    requires_restart: bool = False


class AuthConfigCreate(AuthConfigBase):
    """Schema for creating a new authentication configuration."""


class AuthConfigUpdate(BaseModel):
    """Schema for updating an authentication configuration."""

    config_value: Optional[str] = None
    description: Optional[str] = None


class AuthConfigResponse(AuthConfigBase):
    """Schema for authentication configuration response."""

    id: int
    uuid: str
    created_at: datetime
    updated_at: datetime

    class Config:
        """Pydantic configuration."""

        from_attributes = True


class AuthConfigAuditResponse(BaseModel):
    """Schema for authentication configuration audit log response."""

    id: int
    uuid: str
    config_key: str
    old_value: Optional[str] = None  # Will be masked for sensitive
    new_value: Optional[str] = None  # Will be masked for sensitive
    change_type: str
    ip_address: Optional[str] = None
    created_at: datetime

    class Config:
        """Pydantic configuration."""

        from_attributes = True


# Category-specific configuration schemas


class LDAPConfig(BaseModel):
    """LDAP/Active Directory configuration."""

    ldap_enabled: bool = False
    ldap_server: str = ""
    ldap_port: int = 636
    ldap_use_ssl: bool = True
    ldap_use_tls: bool = False
    ldap_bind_dn: str = ""
    ldap_bind_password: Optional[str] = None  # Sensitive
    ldap_search_base: str = ""
    ldap_username_attr: str = "sAMAccountName"
    ldap_email_attr: str = "mail"
    ldap_name_attr: str = "cn"
    ldap_timeout: int = 10
    ldap_admin_users: str = ""
    ldap_admin_groups: str = ""
    ldap_user_groups: str = ""
    ldap_recursive_groups: bool = False
    ldap_group_attr: str = "memberOf"


class KeycloakConfig(BaseModel):
    """Keycloak/OIDC configuration."""

    keycloak_enabled: bool = False
    keycloak_server_url: str = ""
    keycloak_internal_url: str = ""
    keycloak_realm: str = "opentranscribe"
    keycloak_client_id: str = ""
    keycloak_client_secret: Optional[str] = None  # Sensitive
    keycloak_callback_url: str = ""
    keycloak_admin_role: str = "admin"
    keycloak_timeout: int = 30
    keycloak_verify_audience: bool = True
    keycloak_audience: str = ""
    keycloak_use_pkce: bool = True
    keycloak_verify_issuer: bool = True


class PKIConfig(BaseModel):
    """PKI/X.509 certificate configuration."""

    pki_enabled: bool = False
    pki_ca_cert_path: str = ""
    pki_verify_revocation: bool = False
    pki_cert_header: str = "X-Client-Cert"
    pki_cert_dn_header: str = "X-Client-Cert-DN"
    pki_admin_dns: str = ""
    pki_ocsp_timeout_seconds: int = 5
    pki_crl_cache_seconds: int = 3600
    pki_revocation_soft_fail: bool = True
    pki_trusted_proxies: str = ""
    pki_mode: str = "direct"  # direct, keycloak, hybrid
    pki_allow_password_fallback: bool = True
    pki_support_cac: bool = True
    pki_support_piv: bool = True


class PasswordPolicyConfig(BaseModel):
    """Password policy configuration."""

    password_policy_enabled: bool = True
    password_min_length: int = 12
    password_require_uppercase: bool = True
    password_require_lowercase: bool = True
    password_require_digit: bool = True
    password_require_special: bool = True
    password_history_count: int = 24
    password_max_age_days: int = 60


class MFAConfig(BaseModel):
    """Multi-factor authentication configuration."""

    mfa_enabled: bool = False
    mfa_required: bool = False
    mfa_issuer_name: str = "OpenTranscribe"
    mfa_backup_code_count: int = 10
    mfa_token_expire_minutes: int = 5


class SessionConfig(BaseModel):
    """Session and token configuration."""

    jwt_access_token_expire_minutes: int = 60
    jwt_refresh_token_expire_days: int = 7
    session_idle_timeout_minutes: int = 15
    session_absolute_timeout_minutes: int = 480
    max_concurrent_sessions: int = 5
    concurrent_session_policy: str = "terminate_oldest"


class LoginBannerConfig(BaseModel):
    """Login banner configuration (FedRAMP AC-8)."""

    login_banner_enabled: bool = False
    login_banner_text: str = ""
    login_banner_classification: str = "UNCLASSIFIED"


class AuthMethodTestRequest(BaseModel):
    """Request to test authentication method connection."""

    category: str  # ldap or keycloak
    config: dict[str, Any]


class AuthMethodTestResponse(BaseModel):
    """Response from authentication method test."""

    success: bool
    message: str
    details: Optional[dict[str, Any]] = None


class BulkConfigUpdate(BaseModel):
    """Schema for updating multiple configuration values at once."""

    category: str
    config: dict[str, Any]


class AuthConfigCategoryResponse(BaseModel):
    """Response schema for a category of auth configurations."""

    category: str
    configs: dict[str, Any]


class AuthConfigStatusResponse(BaseModel):
    """Response schema for overall auth configuration status."""

    ldap_enabled: bool = False
    keycloak_enabled: bool = False
    pki_enabled: bool = False
    mfa_enabled: bool = False
    password_policy_enabled: bool = True
    login_banner_enabled: bool = False
